wgsl/index.bs

```diff
@@ -7347,13 +7345,344 @@ See [[#expressions]].
 Statements in a [SHORTNAME] program are executed in control flow order.
 See [[#statements]] and [[#function-calls]].
 
-## Uniformity TODO ## {#uniformity}
+## Uniformity ## {#uniformity}
+
+### Terminology and concepts ### {#uniformity-concepts}
+
+The following definitions are merely informative, trying to give an intuition for what the analysis in the next subsection is computing.
+The analysis is what actually defines these concepts, and when a program is valid or breaks the uniformity rules.
+
+Control flow is said to be uniform at a point in the program if all invocations execute in lockstep at that program point.
+An expression is said to be uniform if it is in uniform control flow, and all invocations compute the same value for it.
+A local variable is said to be uniform, if all invocations have the same value in that variable at every program point where it is live.
+
+### Uniformity analysis overview ### {#uniformity-overview}
+
+In this section we specify an analysis that verifies that functions which are only safe to call in uniform control flow (e.g. barriers and derivatives) are only called in such a context.
+
+<div class="note">Note: This analysis has the following desirable properties:
+      - Sound (meaning that it rejects every program that would break the uniformity requirements of builtins)
```

Comment:
are we only talking about builtins here?
Considering this program:
```
fn foo() -> f32 {
  return myTexture.sample(mySampler, 0.0).x;
}
fn bar() {
  if (non_uniform) {
    foo();
  }
}
```
My interpretation is that we are analysing `foo`'s requirement for being called in uniform control flow, and then we can reject the program based on the call sites of `foo`. And this is user-defined, not a builtin.

Reply:
You are correct. I wrote builtin here, because all constraints ultimately flow from some builtin (if there was no call to `sample` in foo, we could call foo anywhere). I am not sure how to phrase it more clearly.